@philhchen I have been working on integrating this on the pydantic side in https://github.com/pydantic/pydantic/pull/5051; this is great work, very much appreciated!

I've got many tests now passing using this branch for `pydantic_core`, including using int literals as the discriminator. (Doesn't show up like that in CI because it needs this branch, but that's okay 😄).

However, I have noticed two weird issues that I'm hoping you'll see faster than me how to resolve.

Here is the test that I have been working against:
https://github.com/pydantic/pydantic/blob/0aa55db15ffd92343ce4ece22ecb3a0d237bef46/tests/test_discrimated_union.py#L332-L370
(I know the test needs modification for the final bit for the string cases, but it's not getting there yet)

* In the case of `TestStrEnum`, I get the following error while trying to parse the "valid" data:

            @classmethod
            def model_validate(cls: type[Model], obj: Any) -> Model:
        >       values, fields_set = cls.__pydantic_validator__.validate_python(obj)
        E       pydantic_core._pydantic_core.ValidationError: 1 validation error for B
        E       sub -> v_b -> m
        E         Input should be a string, not an instance of a subclass of str [type=string_sub_type, input_value=<EnumValue.b: 'v_b'>, input_type=EnumValue]

    It looks like for StrEnum's to work, we need to modify it to allow proper subclasses of `str`. 

* While the test is passing for IntEnum, you'll notice that the error format it is looking for is actually showing the tag values as though they are strings:

        'ctx': {'discriminator': "'m'", 'expected_tags': "'1', '2'", 'tag': '3'},

    It's possible I made a mistake somewhere on the python side, but I _think_ this is something to do with how the error message is being generated.

(Also, I needed to change the default value of `from_attributes` to `True` for the `TaggedUnionValidator`, as @samuelcolvin mentioned above. I'll push a commit to this branch making that change.)

--- a/pydantic_core/core_schema.py
+++ b/pydantic_core/core_schema.py
@@ -1813,6 +1813,7 @@ class TaggedUnionSchema(TypedDict, total=False):
     custom_error_message: str
     custom_error_context: Dict[str, Union[str, int, float]]
     strict: bool
+    from_attributes: bool  # default: True
     ref: str
     metadata: Any
     serialization: SerSchema
@@ -1826,6 +1827,7 @@ def tagged_union_schema(
     custom_error_message: str | None = None,
     custom_error_context: dict[str, int | str | float] | None = None,
     strict: bool | None = None,
+    from_attributes: bool | None = None,
     ref: str | None = None,
     metadata: Any = None,
     serialization: SerSchema | None = None,
@@ -1873,6 +1875,7 @@ def tagged_union_schema(
         custom_error_message: The custom error message to use if the validation fails
         custom_error_context: The custom error context to use if the validation fails
         strict: Whether the underlying schemas should be validated with strict mode
+        from_attributes: Whether to use the attributes of the object to retrieve the discriminator value
         ref: See [TODO] for details
         metadata: See [TODO] for details
         serialization: Custom serialization schema
@@ -1885,6 +1888,7 @@ def tagged_union_schema(
         custom_error_message=custom_error_message,
         custom_error_context=custom_error_context,
         strict=strict,
+        from_attributes=from_attributes,
         ref=ref,
         metadata=metadata,
         serialization=serialization,

--- a/src/validators/union.rs
+++ b/src/validators/union.rs
@@ -203,7 +203,7 @@ impl ChoiceKey {
         } else if let Ok(py_str) = raw.downcast::<PyString>() {
             Ok(Self::Str(py_str.to_str()?.to_string()))
         } else {
-            py_err!(format!("Expected int or str, got {}", raw.get_type().name()?))
+            py_err!(PyTypeError; "Expected int or str, got {}", raw.get_type().name().unwrap_or("<unknown python object>"))
         }
     }
 }

--- a/src/validators/union.rs
+++ b/src/validators/union.rs
@@ -2,6 +2,7 @@ use std::borrow::Cow;
 use std::fmt;
 use std::fmt::Write;
 
+use pyo3::exceptions::PyTypeError;
 use pyo3::intern;
 use pyo3::prelude::*;
 use pyo3::types::{PyDict, PyList, PyString};
@@ -250,7 +251,7 @@ impl BuildValidator for TaggedUnionValidator {
         let mut descr = String::with_capacity(50);
 
         for (key, value) in schema_choices {
-            let tag = choice_key ChoiceKey::from_py(key)?;
+            let tag = ChoiceKey::from_py(key)?;
 
             if let Ok(repeat_tag) = ChoiceKey::from_py(value) {
                 repeat_choices_vec.push((tag, repeat_tag));
@@ -333,18 +334,10 @@ impl Validator for TaggedUnionValidator {
                         // errors when getting attributes which should be "raised"
                         match lookup_key.$get_method($( $dict ),+)? {
                             Some((_, value)) => {
-                                if self.strict {
-                                    if let Ok(val_int) = value.strict_int() {
-                                        Ok(ChoiceKey::Int(val_int))
-                                    } else {
-                                        Ok(ChoiceKey::Str(value.strict_str()?.as_cow()?.as_ref().to_string()))
-                                    }
+                                if let Ok(int) = value.validate_int(self.strict) {
+                                    Ok(ChoiceKey::Int(int))
                                 } else {
-                                    if let Ok(val_int) = value.lax_int() {
-                                        Ok(ChoiceKey::Int(val_int))
-                                    } else {
-                                        Ok(ChoiceKey::Str(value.lax_str()?.as_cow()?.as_ref().to_string()))
-                                    }
+                                    Ok(ChoiceKey::Str(value.validate_str(self.strict)?.as_cow()?.as_ref().to_string()))
                                 }
                             }
                             None => Err(self.tag_not_found(input)),
@@ -471,7 +464,7 @@ impl TaggedUnionValidator {
                 return match validator.validate(py, input, extra, slots, recursion_guard) {
                     Ok(res) => Ok(res),
                     Err(err) => match tag.as_ref() {
-                        ChoiceKey::Str(s) => Err(err.with_outer_location(s.into())),
+                        ChoiceKey::Str(s) => Err(err.with_outer_location((*s).clone().into())),
                         ChoiceKey::Int(i) => Err(err.with_outer_location((*i).clone().to_string().into())),
                     },
                 };

--- a/tests/validators/test_tagged_union.py
+++ b/tests/validators/test_tagged_union.py
@@ -82,6 +82,7 @@ def test_simple_tagged_union(py_and_json: PyAndJson, input_value, expected):
         {
             'type': 'tagged-union',
             'discriminator': 'foo',
+            'from_attributes': False,
             'choices': {
                 'apple': {
                     'type': 'typed-dict',

